### PR TITLE
introduce goto_programt::instructiont::type() and _nonconst()

### PR DESCRIPTION
**Pull request by kroening**
This introduces two methods to read and write the type of a GOTO program
instruction, and replaces various direct accesses.

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/goto-instrument/uninitialized.cpp [hunk 1]
```diff
@@ -120,11 +120,11 @@ void uninitializedt::add_assertions(
           id2string(identifier)+"#initialized";
 
         symbol_exprt symbol_expr(new_identifier, bool_typet());
-        i1->type=DECL;
+        i1->type_nonconst() = DECL;
         i1->source_location_nonconst() = instruction.source_location();
         i1->code_nonconst() = code_declt(symbol_expr);
 
-        i2->type=ASSIGN;
+        i2->type_nonconst() = ASSIGN;
```
Comment (peterschrammel) [inline src/goto-instrument/uninitialized.cpp]: use the 2-param variant of `insert_after`?

File: src/goto-instrument/uninitialized.cpp [hunk 2]
```diff
@@ -120,11 +120,11 @@ void uninitializedt::add_assertions(
           id2string(identifier)+"#initialized";
 
         symbol_exprt symbol_expr(new_identifier, bool_typet());
-        i1->type=DECL;
+        i1->type_nonconst() = DECL;
```
Comment (peterschrammel) [inline src/goto-instrument/uninitialized.cpp]: use the 2-param variant of `insert_after`?

File: jbmc/src/java_bytecode/remove_exceptions.cpp [hunk 3]
```diff
@@ -421,7 +421,7 @@ bool remove_exceptionst::instrument_throw(
     exc_thrown,
     typecast_exprt(exc_expr, exc_thrown.type()));
   // now turn the `throw' into `assignment'
-  instr_it->type=ASSIGN;
+  instr_it->type_nonconst() = ASSIGN;
```
Comment (peterschrammel) [inline jbmc/src/java_bytecode/remove_exceptions.cpp]: swap?

File: src/goto-instrument/accelerate/disjunctive_polynomial_acceleration.cpp [hunk 4]
```diff
@@ -846,7 +846,7 @@ void disjunctive_polynomial_accelerationt::build_fixed()
   {
     if(instruction.is_assert())
     {
-      instruction.type = ASSUME;
+      instruction.type_nonconst() = ASSUME;
```
Comment (peterschrammel) [inline src/goto-instrument/accelerate/disjunctive_polynomial_acceleration.cpp]: turn_into_assume?

**Review by peterschrammel — COMMENTED** [review]
💡 I think by introducing `turn_into_assume` (we've already got `turn_into_skip`), we can do without `type_nonconst()`.

File: src/goto-programs/goto_program.h [hunk 5]
```diff
@@ -340,8 +340,24 @@ class goto_programt
...
-    goto_program_instruction_typet type;
+    goto_program_instruction_typet type() const
+    {
+      return _type;
+    }
+
+    /// Set the kind of the instruction.
+    /// This method is best avoided to prevent mal-formed instructions.
+    /// Consider using the goto_programt::make_X methods instead.
+    goto_program_instruction_typet &type_nonconst()
```
Comment (martin-cs) [inline src/goto-programs/goto_program.h:351]: As this only has 7 users, would it be better to just refactor those and not have this functionality at all?  Alternatively, could this be marked as deprecated?

**Review by martin-cs — APPROVED** [review]
Seems reasonable.

**Comment by kroening** [comment]
I've added `turn_into_assume`, which removes some usage of `type_nonconst()`.

File: unit/goto-programs/structured_trace_util.cpp [hunk 6]
```diff
@@ -56,7 +56,7 @@ TEST_CASE("structured_trace_util", "[core][util][trace]")
   add_instruction(loop_head_location, instructions);
   // 2: goto 1 # back_edge
   const auto back_edge = add_instruction(back_edge_location, instructions);
-  back_edge->type = GOTO;
+  back_edge->type_nonconst() = GOTO;
```
Comment (tautschnig) [inline unit/goto-programs/structured_trace_util.cpp:59]: Can't we use `make_goto` for this one?

File: src/goto-programs/goto_program.h (hunk 5 above)
Comment (tautschnig) [reply src/goto-programs/goto_program.h:351]: Aren't we even down to 2 users, one of which can easily be fixed and the other being `read_bin_goto_object`? I think the latter could be handled by some low-level constructor that likely wants to be `private` with access only permitted via a `friend` declaration.

**Review by tautschnig — COMMENTED** [review]
It would be great to figure out a story for the `nonconst` case right away.

File: unit/goto-programs/structured_trace_util.cpp (hunk 6 above)
Comment (kroening) [reply unit/goto-programs/structured_trace_util.cpp:59]: add_instruction() in that unit test should be replaced altogether, but this goes beyond this PR.

**Comment by kroening** [comment]
Changing `read_bin_goto_object` is a major rejfactoring of that piece of code; much more involved than the initial scope of the PR. I'd do this separately.

**Review by tautschnig — APPROVED** [review]

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**